Core/Transaction_Handler.php

```diff
@@ -0,0 +1,215 @@
+        $oConfig = $oPayment->getConfig();
+        $oOrder = $oParentTransaction->getTransactionOrder();
+
+        $oTransaction = $this->_getTransactionObject($oPayment, $sActionTitle);
+
+        if (empty($oTransaction)) {
+            $this->_oLogger->error("action not implemented", ['no implementation for ' . $sActionTitle]);
+            return self::_error('Action "' . $sActionTitle . '" is not implemented');
+        }
+
+        $sParentTransactionId = $oParentTransaction->wdoxidee_ordertransactions__transactionid->value;
+        $oTransaction->setParentTransactionId($sParentTransactionId);
+
+        if (!is_null($fAmount)) {
+            $sCurrencyName = $oOrder->oxorder__oxcurrency->value;
+            $oCurrency = Registry::getConfig()->getCurrencyObject($sCurrencyName);
+
+            $oTransaction->setAmount(new Amount(
+                Registry::getUtils()->fRound($fAmount, $oCurrency),
+                $sCurrencyName
+            ));
+        }
+
+        // use the backend service to process the action
+        $oBackendService = new BackendService($oConfig, $this->_oLogger);
+        $oResponse = $oBackendService->process($oTransaction, $sActionTitle);
+        $this->_oLogger->debug('response from ' . $sActionTitle . ' action: ' . $oResponse->getRawData());
+
+        return $this->_handleProcessActionResponse($oResponse, $oOrder, $oBackendService);
+    }
+
+    /**
+     * @param Response       $oResponse
+     * @param Order          $oOrder
+     * @param BackendService $oBackendService
+     *
+     * @return array either [status => success] or [status => error, message => errorMessage]
+     */
+    private function _handleProcessActionResponse($oResponse, $oOrder, $oBackendService)
+    {
+        if ($oResponse instanceof FailureResponse) {
+            $sErrors = '';
+
+            foreach ($oResponse->getStatusCollection()->getIterator() as $oItem) {
+                $sErrors .= $oItem->getDescription() . "<br>";
+                $this->_oLogger->error($oItem->getDescription());
+            }
+
+            return self::_error(Helper::translate('error_transaction') . '<br>' . $sErrors);
+        }
+
+        if ($oResponse instanceof SuccessResponse) {
+            // create new transaction entry in the database
+            $this->_handleNotificationSuccess($oResponse);
+
+            // also update the order state in the database
+            $sUpdatedOrderState = $oBackendService->getOrderState($oResponse->getTransactionType());
+            $oOrder->oxorder__wdoxidee_orderstate = new Field($sUpdatedOrderState);
+            $oOrder->save();
+        }
+
+        return self::_success();
+    }
+
+    /**
+     * @param Payment_Method $oPayment
+     * @param string         $sActionTitle
+     *
+     * @return Transaction
+     */
+    private function _getTransactionObject($oPayment, $sActionTitle)
+    {
+        switch ($sActionTitle) {
+            case 'cancel':
```

Comment:
Please use the `Operations` const values here

Reply:
👍